dataset/src/main/scala/frameless/RecordEncoder.scala

```diff
@@ -44,10 +46,70 @@ object RecordEncoderFields {
     }
 }
 
-class RecordEncoder[F, G <: HList]
+trait NewInstanceExprs[T <: HList] extends Serializable {
+  def from(exprs: List[Expression]): Seq[Expression]
+}
+
+object NewInstanceExprs {
+
+  implicit def deriveHNil: NewInstanceExprs[HNil] = new NewInstanceExprs[HNil] {
+    def from(exprs: List[Expression]): Seq[Expression] = Nil
+  }
+
+  implicit def deriveUnit[K <: Symbol, T <: HList]
+    (implicit
+      tail: NewInstanceExprs[T]
+    ): NewInstanceExprs[FieldType[K, Unit] :: T] = new NewInstanceExprs[FieldType[K, Unit] :: T] {
+      def from(exprs: List[Expression]): Seq[Expression] =
+        Literal.fromObject(()) +: tail.from(exprs)
+    }
+
+  implicit def deriveNonUnit[K <: Symbol, V , T <: HList]
+    (implicit
+      notUnit: V =:!= Unit,
+      tail: NewInstanceExprs[T]
+    ): NewInstanceExprs[FieldType[K, V] :: T] = new NewInstanceExprs[FieldType[K, V] :: T] {
+      def from(exprs: List[Expression]): Seq[Expression] = exprs.head +: tail.from(exprs.tail)
+    }
+}
+
+trait DropUnitValues[L <: HList] extends DepFn1[L] with Serializable { type Out <: HList }
```

Comment:
Unfortunately [filtering in shapeless has performance problems](https://github.com/milessabin/shapeless/issues/596), we made our own simplified filtering implementation for unit values to speed up compilation.